README.md
# DummyText (Chrome Extension)

Simple, quick and unobtrusive dummy text generator for web designers and developers. This extension installs a small icon in your browser allowing you quick access to paragraphs of dummy text (lorem ipsum style) at the click of a button.

The number of paragraphs, words per paragraph and whether or not to wrap paragraphs in paragraph tags can be changed easily. Once happy with your selection, simply click the Copy button and the text is sent straight to your clipboard ready for pasting!

## Installation

Install via the [Chrome Web Store listing](https://chrome.google.com/webstore/detail/dummytext/ahnionhaccolmddlboaklelpadaeafbj).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2.0.0] - 2023-12-23
### Added

- An auto-copy option with a configurable delay
- Support for matching browser colour scheme
- A theme selector with light and dark options
- An options page for configuring settings

### Changed

- Update to Chrome manifest v3
- Rewrote using plain JavaScript
- Split word list into a separate file with the intention to support more in future

## [1.5.0] - 2012-08-04

### Added

- Added README file

## [1.4.0] - 2012-07-07

### Changed

- Set `offline_enabled` setting to true in the manifest file to specify that this extension can run offline.

## [1.3.0] - 2011-11-12

### Changed

- Updated the paragraph checkbox style to look more like Chrome's UI.

## [1.2.0] - 2011-08-11

### Added

- Added a few additional icons.

## [1.1.0] - 2011-08-06

### Changed

- Updated the UI to make it look a little more at home in Chrome.

## [1.0.0] - 2011-08-06

### Added

- Initial release

[unreleased]: https://github.com/davgothic/DummyText/compare/v2.0.0...HEAD
[2.0.0]: https://github.com/davgothic/DummyText/compare/v1.5.0...v2.0.0
[1.5.0]: https://github.com/davgothic/DummyText/compare/v1.4.0...v1.5.0
[1.4.0]: https://github.com/davgothic/DummyText/compare/v1.3.0...v1.4.0
[1.3.0]: https://github.com/davgothic/DummyText/compare/v1.2.0...v1.3.0
[1.2.0]: https://github.com/davgothic/DummyText/compare/v1.1.0...v1.2.0
[1.1.0]: https://github.com/davgothic/DummyText/compare/v1.0.0...v1.1.0
[1.0.0]: https://github.com/davgothic/DummyText/releases/tag/v1.0.0


popup.js
import { load, save } from './settings';

document.addEventListener('DOMContentLoaded', async () => {
  const options = await load();

  document.documentElement.setAttribute('class', `theme-${options.theme}`);

  const response = await fetch('wordlist/lipsum.json');
  const dummyWords = await response.json();
  const punctuation = ['!', '?', '.'];

  const copyButton = document.getElementById('copy-button');
  const lipsumTextArea = document.getElementById('lipsum-textarea');
  const paragraphsInput = document.getElementById('paragraphs-input');
  const wordsInput = document.getElementById('words-input');
  const paragraphTagsCheckbox = document.getElementById('add-tags-checkbox');
  let autoCopyTimeout;

  /**
   * @param {number} max
   * @returns number
   */
  const randomInt = (max) => Math.floor(Math.random() * max);

  /**
   * @param {string[]} words
   * @returns string
   */
  const randomWord = (words) => words[randomInt(words.length)];

  /**
   * @param {number} wordIndex
   * @returns string
   */
  const randomPunctuation = (wordIndex) => {
    if (wordIndex > 1) {
      if (wordIndex % randomInt(30) === 0) {
        return `${randomWord(punctuation)} `;
      }

      if (wordIndex % randomInt(20) === 0) {
        return ', ';
      }
    }
    return ' ';
  };

  /**
   * @param {string} text
   * @returns string
   */
  const capitalizeFirst = (text) => text.replace(/(?:[\n|>|?|!|.]|^)[\s]?([a-z])/g, (s) => s.toUpperCase());

  const generateDummyText = () => {
    clearTimeout(autoCopyTimeout);
    let text = '<p>';

    for (let i = 1; i <= options.paragraphCount; i += 1) {
      for (let j = 1; j <= options.wordsPerParagraph; j += 1) {
        text += randomWord(dummyWords);
        if (j % options.wordsPerParagraph === 0) {
          text += '.</p>\n\n<p>';
        } else if (text.substring(text.length - 3, text.length) !== '<p>') {
          text += randomPunctuation(j);
        }
      }
    }

    text += '</p>';

    if (!options.includeParagraphTags) {
      text = text.replace(/<\/?p>/g, '');
    }

    text = capitalizeFirst(text.substring(0, text.length - (options.includeParagraphTags ? 9 : 2)));

    lipsumTextArea.value = text;
  };

  const copyText = async () => {
    try {
      await navigator.clipboard.writeText(lipsumTextArea.value);
      copyButton.value = 'Copied!';
      copyButton.disabled = 'disabled';
      setTimeout(() => {
        copyButton.value = 'Copy';
        copyButton.disabled = false;
      }, 1000);
    } catch (err) {
      console.error('Failed to copy text.', err.message);
    }
  };

  const saveConfig = async () => {
    if (/\d+/.test(paragraphsInput.value)) {
      if (paragraphsInput.value > 999) {
        paragraphsInput.value = 99;
      } else if (paragraphsInput.value < 1) {
        paragraphsInput.value = 1;
      }
    } else {
      paragraphsInput.value = 1;
    }

    if (/\d+/.test(wordsInput.value)) {
      if (wordsInput.value > 999) {
        wordsInput.value = 999;
      } else if (wordsInput.value < 1) {
        wordsInput.value = 1;
      }
    } else {
      wordsInput.value = 1;
    }

    if (paragraphsInput.value * wordsInput.value > 15000) {
      wordsInput.value = 1;
      paragraphsInput.value = 1;
    }

    options.paragraphCount = parseInt(paragraphsInput.value, 10);
    options.wordsPerParagraph = parseInt(wordsInput.value, 10);
    options.includeParagraphTags = paragraphTagsCheckbox.checked;

    save(options);

    generateDummyText();

    if (options.autoCopy) {
      autoCopyTimeout = setTimeout(copyText, options.autoCopyDelay);
    }
  };

  copyButton.addEventListener('click', copyText);

  paragraphsInput.value = options.paragraphCount;
  paragraphsInput.addEventListener('input', saveConfig);
  paragraphsInput.addEventListener('click', () => paragraphsInput.select());

  wordsInput.value = options.wordsPerParagraph;
  wordsInput.addEventListener('input', saveConfig);
  wordsInput.addEventListener('click', () => wordsInput.select());

  paragraphTagsCheckbox.checked = options.includeParagraphTags;
  paragraphTagsCheckbox.addEventListener('click', saveConfig);

  generateDummyText();
});


options.js
import { load, save } from './settings';

document.addEventListener('DOMContentLoaded', async () => {
  const options = await load();

  document.documentElement.setAttribute('class', `theme-${options.theme}`);
  const enableAutoCopyCheckbox = document.getElementById('enable-autocopy-checkbox');
  const autoCopyDelayInput = document.getElementById('autocopy-delay-input');
  const themeSelect = document.getElementById('theme-select');

  enableAutoCopyCheckbox.checked = options.autoCopy;
  autoCopyDelayInput.value = options.autoCopyDelay;
  themeSelect.value = options.theme;

  // Saves options to chrome.storage
  const saveOptions = async () => {
    if (/\d+/.test(autoCopyDelayInput.value)) {
      if (autoCopyDelayInput.value > 2000) {
        autoCopyDelayInput.value = 2000;
      } else if (autoCopyDelayInput.value < 0) {
        autoCopyDelayInput.value = 0;
      }
    } else {
      autoCopyDelayInput.value = 400;
    }

    options.autoCopy = enableAutoCopyCheckbox.checked;
    options.autoCopyDelay = autoCopyDelayInput.value;
    options.theme = themeSelect.value;

    document.documentElement.setAttribute('class', `theme-${options.theme}`);

    await save(options);
  };

  enableAutoCopyCheckbox.addEventListener('click', saveOptions);
  autoCopyDelayInput.addEventListener('input', saveOptions);
  themeSelect.addEventListener('change', saveOptions);
});


manifest.config.js
import { defineManifest } from '@crxjs/vite-plugin';
import packageJson from './package.json';

const { version } = packageJson;

// Convert from Semver (example: 0.1.0-beta6)
const [major, minor, patch, label = '0'] = version
  // can only contain digits, dots, or dash
  .replace(/[^\d.-]+/g, '')
  // split into version parts
  .split(/[.-]/);

export default defineManifest(async (env) => ({
  manifest_version: 3,
  name:
    env.mode === 'staging'
      ? '[INTERNAL] DummyText'
      : 'DummyText',
  // up to four numbers separated by dots
  version: `${major}.${minor}.${patch}.${label}`,
  // semver is OK in 'version_name'
  version_name: version,
  description: 'Generates random dummy text (Lorem Ipsum).',
  icons: {
    16: 'media/icon19.png',
    48: 'media/icon48.png',
    128: 'media/icon128.png',
  },
  action: {
    default_icon: 'media/icon19.png',
    default_popup: 'popup.html',
  },
  offline_enabled: true,
  options_page: 'options.html',
  permissions: [
    'storage',
  ],
}));


media/scripts/lipsum.js
(function() {
  var lipsum;
  var __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };
  lipsum = {
    numParagraphs: !localStorage["numParagraphs"] ? 3 : localStorage["numParagraphs"],
    numWordsPerParagraph: !localStorage["numWordsPerParagraph"] ? 100 : localStorage["numWordsPerParagraph"],
    addParagraphTags: !localStorage["addParagraphTags"] ? "true" : localStorage["addParagraphTags"],
    autocopy: !localStorage["autocopy"] ? "true" : localStorage["autocopy"],
    dummyWords: ["a", "ab", "ac", "accumsan", "accusamus", "accusantium", "ad", "adipisci", "adipiscing", "adipisicing", "aenean", "alias", "aliqua", "aliquam", "aliquet", "aliquid", "aliquip", "amet", "anim", "animi", "ante", "aperiam", "aptent", "architecto", "arcu", "asperiores", "aspernatur", "assumenda", "at", "atque", "auctor", "augue", "aut", "aute", "autem", "beatae", "bibendum", "blandit", "blanditiis", "cillum", "class", "commodi", "commodo", "condimentum", "congue", "consectetuer", "consectetur", "consequat", "consequatur", "consequuntur", "conubia", "convallis", "corporis", "corrupti", "cras", "cubilia", "culpa", "cum", "cumque", "cupidatat", "cupiditate", "curabitur", "curae", "cursus", "dapibus", "debitis", "delectus", "deleniti", "deserunt", "diam", "diamlorem", "dicta", "dictum", "dictumst", "dignissim", "dignissimos", "dis", "distinctio", "do", "dolor", "dolore", "dolorem", "doloremque", "dolores", "doloribus", "dolorum", "donec", "ducimus", "dui", "duis", "ea", "eaque", "earum", "egestas", "eget", "eius", "eiusmod", "eleifend", "elementum", "eligendi", "elit", "enim", "eos", "erat", "eros", "error", "esse", "est", "et", "etiam", "eu", "euismod", "eum", "eveniet", "ex", "excepteur", "excepturi", "exercitation", "exercitationem", "expedita", "explicabo", "facere", "facilis", "facilisi", "facilisis", "fames", "faucibus", "felis", "fermentum", "feugiat", "fringilla", "fuga", "fugiat", "fugit", "fusce", "gravida", "habitant", "habitasse", "hac", "harum", "hendrerit", "hic", "hymenaeos", "iaculis", "id", "illo", "illum", "impedit", "imperdiet", "in", "inceptos", "incididunt", "incidunt", "integer", "interdum", "inventore", "ipsa", "ipsam", "ipsum", "irure", "iste", "itaque", "iure", "iusto", "justo", "labore", "laboriosam", "laboris", "laborum", "lacinia", "lacus", "laoreet", "laudantium", "lectus", "leo", "libero", "ligula", "litora", "lobortis", "lorem", "luctus", "maecenas", "magna", "magnam", "magni", "magnis", "maiores", "malesuada", "massa", "mattis", "mauris", "maxime", "metus", "mi", "minim", "minima", "minus", "modi", "molestiae", "molestias", "molestie", "mollis", "mollit", "mollitia", "montes", "morbi", "mus", "nam", "nascetur", "natoque", "natus", "nec", "necessitatibus", "nemo", "neque", "nesciunt", "netus", "nibh", "nihil", "nisi", "nisl", "nobis", "non", "nonummy", "nostra", "nostrud", "nostrum", "nulla", "nullam", "numquam", "nunc", "occaecat", "occaecati", "odio", "odit", "officia", "officiis", "omnis", "optio", "orci", "ornare", "pariatur", "parturient", "pede", "pellentesque", "penatibus", "per", "perferendis", "perspiciatis", "pharetra", "phasellus", "placeat", "placerat", "platea", "porro", "porta", "porttitor", "possimus", "posuere", "potenti", "praesent", "praesentium", "pretium", "primis", "proident", "proin", "provident", "pulvinar", "purus", "quae", "quaerat", "quam", "quas", "quasi", "qui", "quia", "quibusdam", "quidem", "quis", "quisquam", "quisque", "quo", "quod", "quos", "ratione", "recusandae", "reiciendis", "rem", "repellat", "repellendus", "reprehenderit", "repudiandae", "rerum", "rhoncus", "ridiculus", "risus", "rutrum", "saepe", "sagittis", "sapien", "sapiente", "scelerisque", "sed", "sem", "semper", "senectus", "sequi", "similique", "sint", "sit", "sociis", "sociosqu", "sodales", "sollicitudin", "soluta", "sunt", "suscipit", "suspendisse", "taciti", "tellus", "tempor", "tempora", "tempore", "temporibus", "tempus", "tenetur", "tincidunt", "torquent", "tortor", "totam", "tristique", "turpis", "ullam", "ullamco", "ullamcorper", "ultrices", "ultricies", "unde", "urna", "ut", "varius", "vehicula", "vel", "velit", "venenatis", "veniam", "veritatis", "vero", "vestibulum", "vitae", "vivamus", "viverra", "voluptas", "voluptate", "voluptatem", "voluptates", "voluptatibus", "voluptatum", "volutpat", "vulputate", "wisi"],
    punctuation: ["!", "?", "."],
    init: function() {
      var copyFunction = function () {
        var copyCallback;
        document.getElementById("lipsum").select();
        document.execCommand("Copy");
        document.getElementById("copy").value = "Copied!";
        document.getElementById("copy").disabled = "disabled";
        copyCallback = function () {
          document.getElementById("copy").value = "Copy to clipboard";
          return document.getElementById("copy").disabled = false;
        };
        return setTimeout(copyCallback, 1000);
      };
      document.getElementById("copy").onclick = copyFunction;
      document.getElementById("paragraphs").value = this.numParagraphs;
      document.getElementById("paragraphs").onkeyup = __bind(function() {
        return this.saveConfig();
      }, this);
      document.getElementById("paragraphs").onclick = function() {
        return this.select();
      };
      document.getElementById("words").value = this.numWordsPerParagraph;
      document.getElementById("words").onkeyup = __bind(function() {
        return this.saveConfig();
      }, this);
      document.getElementById("words").onclick = function() {
        return this.select();
      };
      document.getElementById("ptags").checked = this.addParagraphTags === "true";
      document.getElementById("ptags").onclick = __bind(function() {
        return this.saveConfig();
      }, this);
      document.getElementById("acopy").checked = this.autocopy === "true";
      document.getElementById("acopy").onclick = __bind(function() {
        return this.saveConfig();
      }, this);
      const output = this.generateDummyText();
      if (this.autocopy) {
        copyFunction();
      }
      return output;
    },
    generateDummyText: function() {
      var html, i, j, _ref, _ref2;
      html = "<p>";
      for (i = 1, _ref = this.numParagraphs; 1 <= _ref ? i <= _ref : i >= _ref; 1 <= _ref ? i++ : i--) {
        for (j = 1, _ref2 = this.numWordsPerParagraph; 1 <= _ref2 ? j <= _ref2 : j >= _ref2; 1 <= _ref2 ? j++ : j--) {
          html += this.randomWord(this.dummyWords);
          if ((j > 0) && (j % this.numWordsPerParagraph === 0)) {
            html += ".</p>\n\n<p>";
          } else {
            if (html.substring(html.length - 3, html.length) !== "<p>") {
              html += this.randomPunctuation(j);
            }
          }
        }
      }
      html += "</p>";
      if (this.addParagraphTags !== "true") {
        html = html.replace(/<\/?p>/g, "");
      }
      html = this.capitalizeFirst(html.substring(0, html.length - (this.addParagraphTags === "true" ? 9 : 2)));
      return document.getElementById("lipsum").innerHTML = html;
    },
    saveConfig: function() {
      var paragraphs, words;
      words = document.getElementById("words").value;
      paragraphs = document.getElementById("paragraphs").value;
      if (/\d+/.test(paragraphs)) {
        this.numParagraphs = localStorage["numParagraphs"] = paragraphs;
      }
      if (/\d+/.test(words)) {
        this.numWordsPerParagraph = localStorage["numWordsPerParagraph"] = words;
      }
      this.addParagraphTags = localStorage["addParagraphTags"] = document.getElementById("ptags").checked.toString();
      this.autocopy = localStorage["autocopy"] = document.getElementById("acopy").checked.toString();
      return this.generateDummyText();
    },
    randomWord: function(words) {
      return words[this.randomInt(words.length)];
    },
    randomPunctuation: function(x) {
      if (x > 0) {
        if (x % this.randomInt(30) === 0) {
          return this.randomWord(this.punctuation) + " ";
        } else if (x % this.randomInt(20) === 0) {
          return ", ";
        } else {
          return " ";
        }
      }
      return " ";
    },
    randomInt: function(x) {
      return Math.floor(Math.random() * parseInt(x));
    },
    capitalizeFirst: function(x) {
      var er;
      er = /(?:[\n|>|\?|\!|\.]|^)[\s]?([a-z])/g;
      return x.replace(er, function(s) {
        return s.toUpperCase();
      });
    }
  };
  if (typeof window !== "undefined" && window !== null) {
    window.lipsum = lipsum;
  }
  window.onload = function() {
    return lipsum.init();
  };
}).call(this);
